src/audio/aec/aec.c

```diff
@@ -0,0 +1,543 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/aec/aec.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define AEC_BLK_LENGTH 4
+#define AEC_MAX_SOURCES 2 /* Microphone and reference streams */
+
+static const struct comp_driver comp_aec;
+
+/* UUID 2ca424c0-7e1c-4b0a-afca-43de944705d3 */
+DECLARE_SOF_RT_UUID("aec", aec_uuid, 0x2ca424c0, 0x7e1c, 0x4b0a, 0xaf, 0xca,
+		    0x43, 0xde, 0x94, 0x47, 0x05, 0xd3);
+
+DECLARE_TR_CTX(aec_tr, SOF_UUID(aec_uuid), LOG_LEVEL_INFO);
+
+/* Generic processing */
+
+#if CONFIG_FORMAT_S16LE
+
+static void aec_s16_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames)
+
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int64_t mix;
+	int32_t ref;
+	int16_t out;
+	int16_t *x;
+	int16_t *y;
+	int16_t *r16;
+	int32_t *r32;
+	int in_idx;
+	int ref_idx;
+	int ref_nch;
+	int i;
+	int j;
+	int nch = sources[0]->channels;
+	int out_idx = 0;
+
+	if (cd->ref_active)
+		ref_nch = sources[1]->channels;
+
+	/* Input from selected input and reference channels, same output to every channel */
+	in_idx = cd->source_channel;
+	ref_idx = cd->reference_channel;
+	for (i = 0; i < frames; i++) {
+		x = audio_stream_read_frag_s16(sources[0], in_idx);
+		mix = *x << 16;
+		in_idx += nch;
+		if (cd->ref_active) {
+			if (cd->ref_32bits) {
+				r32 = audio_stream_read_frag_s32(sources[1], ref_idx);
+				ref = *r32;
+			} else {
+				r16 = audio_stream_read_frag_s16(sources[1], ref_idx);
+				ref = *r16;
+			}
+			ref_idx += ref_nch;
+			mix = mix - (ref << cd->ref_shift);
+		}
```

Comment:
This ref_active code block is running the same conditional logic within a processing loop. Can we optimise this out and have separate loops depending on ref_active or ref_32bits.

Reply:
Yep. There will be adaptive filter that processes `ref` in next version.